File: holoken-core/src/main/java/com/holokenmod/calculation/GridCalculationService.java
package com.holokenmod.calculation;

import com.holokenmod.creation.GridCalculator;
import com.holokenmod.grid.Grid;
import com.holokenmod.options.GameVariant;

import java.util.ArrayList;
import java.util.Optional;

public class GridCalculationService {
	private static final GridCalculationService INSTANCE = new GridCalculationService();
	
	private final ArrayList<GridCalculationListener> listeners = new ArrayList<>();
	private Optional<Grid> nextGrid = Optional.empty();
	private GameVariant variant;
	
	public static GridCalculationService getInstance() {
		return INSTANCE;
	}
	
	public void addListener(GridCalculationListener listener) {
		listeners.add(listener);
	}
	
	public void calculateCurrentAndNextGrids(GameVariant variant) {
		this.nextGrid = Optional.empty();
		this.variant = variant;
		
		calculateCurrentGrid();
		calculateNextGrid();
	}
	
	private void calculateCurrentGrid() {
		listeners.forEach(GridCalculationListener::startingCurrentGridCalculation);
		
		final GridCalculator creator = new GridCalculator(variant);
		
		Grid newGrid = creator.calculate();
		
		listeners.forEach(listener -> listener.currentGridCalculated(newGrid));
	}
	
	public void calculateNextGrid() {
		listeners.forEach(GridCalculationListener::startingNextGridCalculation);
		
		final GridCalculator creator = new GridCalculator(variant);
		
		Grid grid = creator.calculate();
		
		nextGrid = Optional.of(grid);
		
		listeners.forEach(listener -> listener.nextGridCalculated(grid));
	}
	
	public boolean hasCalculatedNextGrid(GameVariant variantParam) {
		return nextGrid.isPresent()
			&& variantParam.equals(variant);
	}
	
	public Grid consumeNextGrid() {
		Grid grid = nextGrid.get();
		
		nextGrid = Optional.empty();
		
		return grid;
	}
	
	public void setVariant(GameVariant variant) {
		this.variant = variant;
	}
	
	public void setNextGrid(Grid grid) {
		nextGrid = Optional.of(grid);
	}
}

File: holoken-core/src/main/kotlin/com/holokenmod/calculation/GridCalculationService.kt
package com.holokenmod.calculation

import com.holokenmod.creation.GridCalculator
import com.holokenmod.grid.Grid
import com.holokenmod.options.GameVariant
import java.util.function.Consumer

class GridCalculationService(
    private var variant: GameVariant
) {
    private val listeners = mutableListOf<GridCalculationListener>()
    private var nextGrid: Grid? = null

    fun addListener(listener: GridCalculationListener) {
        listeners += listener
    }

    fun calculateCurrentAndNextGrids(variant: GameVariant) {
        nextGrid = null
        this.variant = variant
        calculateCurrentGrid()
        calculateNextGrid()
    }

    private fun calculateCurrentGrid() {
        listeners.forEach(Consumer { obj: GridCalculationListener -> obj.startingCurrentGridCalculation() })
        val creator = GridCalculator(variant)
        val newGrid = creator.calculate()
        listeners.forEach(Consumer { listener: GridCalculationListener ->
            listener.currentGridCalculated(
                newGrid
            )
        })
    }

    fun calculateNextGrid() {
        listeners.forEach(Consumer { obj: GridCalculationListener -> obj.startingNextGridCalculation() })
        val creator = GridCalculator(variant)
        val grid = creator.calculate()
        nextGrid = grid
        listeners.forEach(Consumer { listener: GridCalculationListener ->
            listener.nextGridCalculated(
                grid
            )
        })
    }

    fun hasCalculatedNextGrid(variantParam: GameVariant): Boolean {
        return nextGrid != null && variantParam == variant
    }

    fun consumeNextGrid(): Grid {
        val grid = nextGrid!!
        nextGrid = null
        return grid
    }

    fun setVariant(variant: GameVariant) {
        this.variant = variant
    }

    fun setNextGrid(grid: Grid) {
        nextGrid = grid
    }
}